text/0116-sentry-semantic-conventions.md

```diff
@@ -0,0 +1,159 @@
+
+For replays we have:
+
+- [Top level fields](https://develop.sentry.dev/sdk/event-payloads/) like [user](https://develop.sentry.dev/sdk/event-payloads/user/)/release/environment.
+- Tags, string to string key value pairs that are indexed and searchable in Sentry.
+- Breadcrumbs data, which is heavily relied on by the replay product.
+
+For crons we have:
+
+- [Top level fields](https://develop.sentry.dev/sdk/event-payloads/) like release/environment.
+- [Contexts](https://develop.sentry.dev/sdk/event-payloads/contexts/), our current implementation of arbitrary contextual data attached to events. Only the trace context is supported for crons at the current time.
+
+For metrics we have:
+
+- Tags
+
+Given the above, we can categorize the data attached to Sentry signals as so:
+
+- String key to string value pairs (tags, top level fields)
+- Structured dictionary with well defined keys and values (contexts)
+- Unstructured dictionary with arbitrary keys and values (breadcrumb data, span data)
+
+Currently this is confusing, and it can even be hard to tell what data is generated by an sdk, added by a user, or injected at processing time.
+
+# Sentry Semantic Conventions
+
+This RFC proposes adding semantic conventions that behave exactly like OpenTelemetry's semantic conventions. Each signal will get a new `attributes` field that is dictionary of key value pairs. For purposes of backwards comptability, this field can also be called `data` (means it can be instantly adopted for breadcrumbs and spans), but new signals like crons should use `attributes`.
+
+For the purposes of rollout, we recommend this field is adopted by spans, metrics (DDM), and breadcrumbs first, and then adopted by crons after it has been further validated given there is no arbitrary context on crons at the moment.
+
+This does not replace tags, which should still be part of their respective signal payloads. The product can make decisions on how to promote different attributes to become tags. This happens with contexts and span data today.
+
+## Attributes Schema
+
+Attribute keys should be unique and well-known, and should not be used for multiple purposes. In OpenTelemetry the attribute values cannot be dictionaries, only primitives and arrays of primitives. We will try to follow this convention as well, but we will allow dictionaries as attribute values for better backwards compatibility with existing Sentry fields.
+
+```ts
+export interface Attributes {
+  [attributeKey: string]: AttributeValue;
+}
+
+/**
+ * Attribute values may be any non-nullish primitive value except an object.
+ *
+ * null or undefined attribute values are invalid and will result in undefined behavior.
+ */
+export type AttributeValue =
+  | string
+  | number
+  | boolean
+  | Array<string>
+  | Array<number>
+  | Array<boolean>
```

Comment:
This will mean no mixed arrays? eg `['foo', 42, true]`?

It's probably fine but just raising to be sure somebody thought about this.